File: Packages/Yage/Sources/Yage/Models/Capability.swift
import Schwifty
import SwiftUI

open class Capability {
    public weak var subject: Entity?
    public var isEnabled: Bool = true

    public lazy var tag: String = {
        let name = String(describing: type(of: self))
        let id = subject?.id ?? "n/a"
        return "\(name)-\(id)"
    }()
    
    public required init() {}

    open func install(on subject: Entity) {
        Logger.log(subject.id, "Installing", String(describing: self))
        self.subject = subject
    }
    
    func update(with collisions: Collisions, after time: TimeInterval) {
        doUpdate(with: collisions, after: time)
    }
    
    open func doUpdate(with collisions: Collisions, after time: TimeInterval) {
    }

    open func kill(autoremove: Bool = true) {
        if autoremove {
            subject?.capabilities.removeAll { $0.tag == tag }
        }
        subject = nil
        isEnabled = false
    }
}

public extension Entity {
    func install(_ capability: Capability) {
        capability.install(on: self)
        capabilities.append(capability)
    }
}

open class CapabilitiesDiscoveryService {
    public static var shared = CapabilitiesDiscoveryService()
    
    public init() {}
    
    open func capability(for id: String) -> Capability? {
        fatalError("Override CapabilitiesDiscoveryService.shared with your own instance.")
    }
}

File: Sources/windows/yage/models/capability.py
from typing import Optional

from yage.utils.logger import Logger


class Capability:
    def __init__(self, subject):
        self.subject = subject
        self.is_enabled = True
        self.tag = f'{subject.id}][{type(self).__name__}'
        Logger.log(subject.id, f'Installing {type(self).__name__}')

    def update(self, collisions, time):
        if self.is_enabled:
            self.do_update(collisions, time)

    def do_update(self, collisions, time):
        pass

    def kill(self, autoremove=True):
        if autoremove:
            self._remove_from_subject_capabilities()
        self.subject = None
        self.is_enabled = False

    def _remove_from_subject_capabilities(self):
        self.subject.capabilities = [
            c for c in self.subject.capabilities if c.tag != self.tag]


class CapabilitiesDiscoveryService:
    #pylint: disable=unused-argument
    def capability(self, capability_id: str) -> Optional[Capability]:
        return None
